src/nbt/utils.rs

```diff
@@ -42,3 +44,55 @@ where
         })
         .collect()
 }
+
+// like [T]::join, but allowing for formatting
+// Runs a sequence of formatting functions, interspersed with instances of `separator`
+pub(crate) fn join_formatted<Separator, I, F>
+    (f: &mut Formatter<'_>, separator: Separator, iterator: I) -> std::fmt::Result 
+    where Separator: Clone + Display, 
+            I: IntoIterator<Item = F>, 
+            F: FnOnce(&mut Formatter<'_>) -> std::fmt::Result,
+{
+    let mut peekable = iterator.into_iter().peekable();
+    while let Some(function) = peekable.next() {
+        function(f)?;
+        if peekable.peek().is_some() {
+            write!(f, "{}", separator)?;
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn escape_name(s: &String) -> String {
+    // This is not quite perfect, given that it always iterates s twice
+    // I would wish for a better solution, but it would be quite complicated
+    let may_be_unquoted = !s.is_empty() && s.chars()
+        .all(|c| c.is_alphanumeric() || c == '.' || c == '_' || c == '+' || c == '-' );
+    if may_be_unquoted { s.clone() } 
+    else { escape_string_value(s) }
+}
+
+pub(crate) fn escape_string_value(s: &str) -> String {
+    let mut output = String::with_capacity(s.as_bytes().len() + 2); // +2 because ""
+    // using str here is a bit weird, but it is the best way to allow use of String::replace_range
```

Comment:
For readability and performance reasons, I would prefer to keep it `char` and when using `escape_char` add `.to_string()`